SingularityUI/app/components/newDeployForm/NewDeployForm.jsx

```diff
@@ -0,0 +1,1423 @@
+            }
+            if (artifact.type === 'external') {
+              this.copyFieldsToObject(newArtifact, ARTIFACT_FIELDS.external, (id) => artifact[id] || INDEXED_ALL_ARTIFACT_FIELDS[id].default);
+            }
+            if (artifact.type === 's3') {
+              this.copyFieldsToObject(newArtifact, ARTIFACT_FIELDS.s3, (id) => artifact[id] || INDEXED_ALL_ARTIFACT_FIELDS[id].default);
+            }
+            return newArtifact;
+          });
+          deployObject[fieldId.id] = artifacts;
+        } else if (fieldId.type === 'volumes') {
+          const volumes = value.map(volume => this.copyFieldsToObject(
+            {},
+            DOCKER_VOLUME_FIELDS,
+            (id) => volume[id] || INDEXED_DOCKER_VOLUME_FIELDS[id].default
+          ));
+          deployObject[fieldId.id] = volumes;
+        } else if (fieldId.type === 'portMappings') {
+          const portMappings = value.map(portMapping => this.copyFieldsToObject(
+            {},
+            DOCKER_PORT_MAPPING_FIELDS,
+            (id) => portMapping[id] || INDEXED_DOCKER_PORT_MAPPING_FIELDS[id].default));
+          deployObject[fieldId.id] = portMappings;
+        }
+      }
+    }
+    return deployObject;
+  }
+
+  submit(event) {
+    event.preventDefault();
+    const deployObject = {};
+    this.copyFieldsToObject(deployObject, FIELDS.all, (fieldId) => this.getValueOrDefault(fieldId));
+    if (this.getValueOrDefault('executorType') === DEFAULT_EXECUTOR_TYPE) {
+      this.copyFieldsToObject(deployObject, FIELDS.defaultExecutor, (fieldId) => this.getValueOrDefault(fieldId));
+    } else {
+      this.copyFieldsToObject(deployObject, FIELDS.customExecutor, (fieldId) => this.getValueOrDefault(fieldId));
+    }
+    if (this.getValueOrDefault('type') === 'docker') {
+      this.copyFieldsToObject(deployObject, FIELDS.dockerContainer, (fieldId) => this.getValueOrDefault(fieldId));
+    }
+    if (this.props.request.request.loadBalanced) {
+      this.copyFieldsToObject(deployObject, FIELDS.loadBalancer, (fieldId) => this.getValueOrDefault(fieldId));
+    }
+    if (this.isRequestDaemon()) {
+      this.copyFieldsToObject(deployObject, FIELDS.healthChecker, (fieldId) => this.getValueOrDefault(fieldId));
+    }
+    deployObject.requestId = this.props.request.request.id;
+    this.props.save({deploy: deployObject});
+  }
+
+  addDeployObjectToArrayField(fieldId, deployObject) {
+    if (!this.props.form[fieldId]) {
+      this.updateField(fieldId, [deployObject]);
+    } else {
+      const fieldValue = this.props.form[fieldId].slice();
+      fieldValue.push(deployObject);
+      this.updateField(fieldId, fieldValue);
+    }
+  }
+
+  addObjectToDeployFieldPreventDefault(fieldId, deployObject, event) {
+    event.preventDefault();
+    this.addDeployObjectToArrayField(fieldId, deployObject);
+  }
+
+  removeThingFromArrayField(fieldId, key) {
```

Comment:
So the things that can get added are `artifact`s, `portMapping`s, and `volume`s. I don't necessarily think `deployObject` is the best name for that category of things, but I haven't been able to think of a better one. Maybe `SubField`?